File: Frontend/src/types/contracts/interviews.ts
import { z } from "zod";

import { topicMetricSchema } from "@/types/contracts/analytics";
import { interviewStatusSchema, interviewTypeSchema } from "@/types/contracts/common";
import { preparationTaskSchema } from "@/types/contracts/preparation";

export const interviewRoundSchema = z.object({
  id: z.string(),
  name: z.string(),
  type: interviewTypeSchema,
  status: z.enum(["completed", "current", "pending"]),
});

export const interviewSchema = z.object({
  id: z.string(),
  company: z.string(),
  companyMark: z.string(),
  role: z.string(),
  type: interviewTypeSchema,
  round: z.string(),
  roundNumber: z.number(),
  totalRounds: z.number(),
  scheduledAt: z.string(),
  timezone: z.string(),
  durationMinutes: z.number(),
  meetingUrl: z.string().optional(),
  recruiter: z.string().optional(),
  interviewers: z.array(z.string()).optional(),
  status: interviewStatusSchema,
  readiness: z.number(),
  preparationProgress: z.number(),
  location: z.string(),
  accent: z.string(),
  rounds: z.array(interviewRoundSchema),
});

export const dashboardOverviewSchema = z.object({
  nextInterview: interviewSchema.nullable(),
  upcomingInterviews: z.array(interviewSchema),
  todayTasks: z.array(preparationTaskSchema),
  weakTopics: z.array(topicMetricSchema),
  streakDays: z.number(),
  scoreTrend: z.array(z.number()),
  readinessDeltaThisWeek: z.number(),
});

export const createInterviewRequestSchema = z.object({
  company: z.string(),
  role: z.string(),
  type: interviewTypeSchema,
  scheduledAt: z.string(),
  timezone: z.string(),
});

export const updateInterviewRequestSchema = createInterviewRequestSchema
  .partial()
  .extend({
    meetingUrl: z.string().optional(),
    recruiter: z.string().optional(),
    interviewers: z.array(z.string()).optional(),
  });

File: Backend/app/schemas/interviews.py
from typing import ClassVar

from app.core.serialization import CamelModel
from app.core.timeutils import UtcDatetime
from app.schemas.common import InterviewStatus, InterviewType, RoundStatus


class InterviewRound(CamelModel):
    id: str
    name: str
    type: InterviewType
    status: RoundStatus


class Interview(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"meeting_url", "recruiter", "interviewers"})

    id: str
    company: str
    company_mark: str
    role: str
    type: InterviewType
    round: str
    round_number: int
    total_rounds: int
    scheduled_at: UtcDatetime
    timezone: str
    duration_minutes: int
    meeting_url: str | None = None
    recruiter: str | None = None
    interviewers: list[str] | None = None
    status: InterviewStatus
    readiness: int
    preparation_progress: int
    location: str
    accent: str
    rounds: list[InterviewRound]


class CreateInterviewRequest(CamelModel):
    company: str
    role: str
    type: InterviewType
    scheduled_at: UtcDatetime
    timezone: str


class UpdateInterviewRequest(CamelModel):
    company: str | None = None
    role: str | None = None
    type: InterviewType | None = None
    scheduled_at: UtcDatetime | None = None
    timezone: str | None = None
    meeting_url: str | None = None
    recruiter: str | None = None
    interviewers: list[str] | None = None
